Fix #671 - fix and consider tycho.mode=maven

HannesWell opened the pull request:
This PR aims to fix issue #671 that `tycho.mode=maven` is not considered correctly anymore.

The first problem is that the value of `tycho.mode` is overwritten in `TychoGraphBuilder` and `TychoPomlessLifecycleParticipant` by the value `extension`. Since I was not able to identify a use case for the overwritten value I simply removed the assignment.

The second problem is that the `TargetPlatformMojo` and the `ValidateClasspathMojo` are not executed correctly when `tycho.mode=maven` is set. One way is to handle this in Tycho is to skip the runs as I just did now in this PR. Another way should be to handle this in the m2e build by binding the default execution of those two mojos to the `None` phase and to suppress them this way. I will check if that works out tomorrow.

laeubi commented:
> The first problem is that the value of `tycho.mode` is overwritten in `TychoGraphBuilder` and `TychoPomlessLifecycleParticipant` by the value `extension`. Since I was not able to identify a use case for the overwritten value I simply removed the assignment.

You should either use `tycho-builder` or `tycho.mode=maven` not both. Even though this currently is *sometimes* possible the flag should prevent this so people don't start rely on it.

If this is required for migration, you can use the old pomless-extension for now, but be aware that it might be removed/disabled in Tycho 3.0.

> The second problem is that the `TargetPlatformMojo` and the `ValidateClasspathMojo` are not executed correctly when `tycho.mode=maven` is set.

Can you explain what that means? Is there an exception thrown? If both are expected to be skipped with `tycho.mode=maven` they could simply check this property in the `execute()` method.

HannesWell commented:
> > The first problem is that the value of `tycho.mode` is overwritten in `TychoGraphBuilder` and `TychoPomlessLifecycleParticipant` by the value `extension`. Since I was not able to identify a use case for the overwritten value I simply removed the assignment.
> 
> You should either use `tycho-builder` or `tycho.mode=maven` not both. Even though this currently is _sometimes_ possible the flag should prevent this so people don't start rely on it.

But how am I then supposed to handle situations in M2E where a first build in 'Maven-Mode' is done where I don't want the target-resolution to happen but in a subsequent build it should happen?
I have updated the M2E PR to also update the extensions to how it should be (is that right?), except for the -SNAPSHOT suffix: 
https://github.com/eclipse-m2e/m2e-core/pull/589/files

With Tycho's current 2.7.x branch HEAD installed, when launching the preliminary M2E build with
`mvn clean generate-sources -f m2e-maven-runtime -P generate-osgi-metadata -Dtycho.mode=maven` respectivly m2e's launch config `m2e-maven-runtime--generate-OSGi-metadata`, I get the following failing build:

```
[INFO] Scanning for projects...
[ERROR] Internal error: java.lang.RuntimeException: Could not resolve target platform specification artifact org.eclipse.m2e:target-platform:target:1.16.0-SNAPSHOT -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Could not resolve target platform specification artifact org.eclipse.m2e:target-platform:target:1.16.0-SNAPSHOT
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:120)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:972)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:293)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:196)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
Caused by: java.lang.RuntimeException: Could not resolve target platform specification artifact org.eclipse.m2e:target-platform:target:1.16.0-SNAPSHOT
	at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.addTargetArtifact(DefaultTargetPlatformConfigurationReader.java:459)
	at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.setTarget(DefaultTargetPlatformConfigurationReader.java:372)
	at org.eclipse.tycho.core.resolver.DefaultTargetPlatformConfigurationReader.getTargetPlatformConfiguration(DefaultTargetPlatformConfigurationReader.java:102)
	at org.eclipse.tycho.core.resolver.DefaultTychoResolver.setupProject(DefaultTychoResolver.java:102)
	at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead(TychoMavenLifecycleParticipant.java:111)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:264)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	... 11 more
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

The cause of this failure is that the `TychoGraphBuilder` and the `TychoPomlessLifecycleParticipant` overwrite the value of the "tycho.mode" property to "extension". Subsequently `TychoMavenLifecycleParticipant.disableLifecycleParticipation()` returns false and TychoMavenLifecycleParticipant.afterProjectsRead() executes normally which leads to the error above. And I know no other way than `tycho.mode=maven` to suppress that. 

While the first overwrite is expected, having `TychoPomlessLifecycleParticipant` in the build is surprising to me because `tycho-build` is actually configured and not `tycho-pomless`. @laeubi do you have any idea why both kick in?

Nevertheless I could 'workaround' the problem by completing my works towards a single-shot build of M2E that we started about new-year in m2e (Sorry about delaying that). But in the meantime there have to be means to disable the initial TP resolution.

> 
> If this is required for migration, you can use the old pomless-extension for now, but be aware that it might be removed/disabled in Tycho 3.0.

Actually I would like to avoid the now 'legacy' extension tycho-pomless and move forward.

> 
> > The second problem is that the `TargetPlatformMojo` and the `ValidateClasspathMojo` are not executed correctly when `tycho.mode=maven` is set.
> 
> Can you explain what that means? Is there an exception thrown?

Right they throw exceptions.
First 

```
[INFO] --- target-platform-configuration:2.7.0-SNAPSHOT:target-platform (default-target-platform) @ org.eclipse.m2e.archetype.common ---
[DEBUG] Dependency collection stats {ConflictMarker.analyzeTime=35200, ConflictMarker.markTime=20600, ConflictMarker.nodeCount=27, ConflictIdSorter.graphTime=8000, ConflictIdSorter.topsortTime=12700, ConflictIdSorter.conflictIdCount=23, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=178200, ConflictResolver.conflictItemCount=27, DefaultDependencyCollector.collectTime=1584700, DefaultDependencyCollector.transformTime=267600}
[DEBUG] org.eclipse.tycho:target-platform-configuration:jar:2.7.0-SNAPSHOT
[DEBUG]    org.eclipse.tycho:sisu-equinox-api:jar:2.7.0-SNAPSHOT:compile
[DEBUG]    org.eclipse.tycho:org.eclipse.tycho.p2.resolver.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]    org.eclipse.tycho:tycho-core:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:org.eclipse.tycho.p2.tools.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.codehaus.plexus:plexus-archiver:jar:4.2.7:compile (version managed from default)
[DEBUG]          org.codehaus.plexus:plexus-utils:jar:3.4.1:compile (version managed from default)
[DEBUG]          org.codehaus.plexus:plexus-io:jar:3.2.0:compile (version managed from default)
[DEBUG]             commons-io:commons-io:jar:2.6:compile
[DEBUG]          org.apache.commons:commons-compress:jar:1.21:compile (version managed from default)
[DEBUG]          org.iq80.snappy:snappy:jar:0.4:compile
[DEBUG]          org.tukaani:xz:jar:1.9:runtime
[DEBUG]       org.codehaus.plexus:plexus-interpolation:jar:1.26:compile
[DEBUG]       org.codehaus.plexus:plexus-cipher:jar:2.0:compile (version managed from default)
[DEBUG]          javax.inject:javax.inject:jar:1:compile
[DEBUG]       org.eclipse.tycho:sisu-equinox-embedder:jar:2.7.0-SNAPSHOT:compile
[DEBUG]          org.codehaus.plexus:plexus-component-annotations:jar:2.1.1:compile (version managed from default)
[DEBUG]          org.eclipse.platform:org.eclipse.osgi:jar:3.17.100:compile (version managed from default)
[DEBUG]       org.eclipse.tycho:tycho-metadata-model:jar:2.7.0-SNAPSHOT:compile
[DEBUG]          de.pdark:decentxml:jar:1.4:compile (version managed from default)
[DEBUG]       org.eclipse.tycho:tycho-embedder-api:jar:2.7.0-SNAPSHOT:compile
[DEBUG]          org.eclipse.tycho:org.eclipse.tycho.embedder.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:org.eclipse.tycho.core.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG] Created new class realm plugin>org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT
[DEBUG] Importing foreign packages into class realm plugin>org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT
[DEBUG]   Imported:  < project>org.eclipse.m2e:m2e-maven-runtime:1.16.0-SNAPSHOT
[DEBUG] Populating class realm plugin>org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT
[DEBUG]   Included: org.eclipse.tycho:target-platform-configuration:jar:2.7.0-SNAPSHOT
[DEBUG]   Included: org.codehaus.plexus:plexus-archiver:jar:4.2.7
[DEBUG]   Included: org.codehaus.plexus:plexus-utils:jar:3.4.1
[DEBUG]   Included: org.codehaus.plexus:plexus-io:jar:3.2.0
[DEBUG]   Included: commons-io:commons-io:jar:2.6
[DEBUG]   Included: org.apache.commons:commons-compress:jar:1.21
[DEBUG]   Included: org.iq80.snappy:snappy:jar:0.4
[DEBUG]   Included: org.tukaani:xz:jar:1.9
[DEBUG]   Included: org.codehaus.plexus:plexus-interpolation:jar:1.26
[DEBUG]   Included: org.codehaus.plexus:plexus-cipher:jar:2.0
[DEBUG]   Included: org.codehaus.plexus:plexus-component-annotations:jar:2.1.1
[DEBUG]   Included: org.eclipse.platform:org.eclipse.osgi:jar:3.17.100
[DEBUG] Configuring mojo org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform from plugin realm ClassRealm[plugin>org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@3764951d]
[DEBUG] Configuring mojo 'org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform' with basic configurator -->
[DEBUG]   (f) project = MavenProject: org.eclipse.m2e:org.eclipse.m2e.archetype.common:1.18.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\m2e-maven-runtime\org.eclipse.m2e.archetype.common\pom.xml
[DEBUG] -- end configuration --
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-bundles-external:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-bundles-external:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Extracting Tycho's OSGi runtime
[DEBUG] Skipped remote request for org.eclipse.tycho:org.eclipse.tycho.p2.resolver.impl:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:org.eclipse.tycho.p2.resolver.impl:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:org.eclipse.tycho.p2.maven.repository:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:org.eclipse.tycho.p2.maven.repository:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:org.eclipse.tycho.p2.tools.impl:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:org.eclipse.tycho.p2.tools.impl:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
Debug options:
    file:/C:/dev/git/eclipse.m2e-core/.options not found
Time to load bundles: 170
[DEBUG] Clearing proxy settings in OSGi runtime
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] M2E - Maven runtime bundles 1.16.0-SNAPSHOT ........ SUCCESS [  1.119 s]
[INFO] M2E Maven Archetype Common 1.18.1-SNAPSHOT ......... FAILURE [  2.151 s]
[INFO] M2E Maven/Nexus Indexer 1.18.2-SNAPSHOT ............ SKIPPED
[INFO] M2E Embedded Maven Runtime (includes Incubating components) 1.18.2-SNAPSHOT SKIPPED
[INFO] M2E SLF4J-Simple Binding for Embedded Maven Runtime 1.18.1-SNAPSHOT SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.857 s
[INFO] Finished at: 2022-02-22T09:36:59+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform (default-target-platform) on project org.eclipse.m2e.archetype.common: Execution default-target-platform of goal org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform failed: Tycho build extension not configured for MavenProject: org.eclipse.m2e:org.eclipse.m2e.archetype.common:1.18.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\m2e-maven-runtime\org.eclipse.m2e.archetype.common\pom.xml -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform (default-target-platform) on project org.eclipse.m2e.archetype.common: Execution default-target-platform of goal org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform failed: Tycho build extension not configured for MavenProject: org.eclipse.m2e:org.eclipse.m2e.archetype.common:1.18.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\m2e-maven-runtime\org.eclipse.m2e.archetype.common\pom.xml
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:972)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:293)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:196)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-target-platform of goal org.eclipse.tycho:target-platform-configuration:2.7.0-SNAPSHOT:target-platform failed: Tycho build extension not configured for MavenProject: org.eclipse.m2e:org.eclipse.m2e.archetype.common:1.18.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\m2e-maven-runtime\org.eclipse.m2e.archetype.common\pom.xml
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:148)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	... 20 more
Caused by: java.lang.IllegalStateException: Tycho build extension not configured for MavenProject: org.eclipse.m2e:org.eclipse.m2e.archetype.common:1.18.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\m2e-maven-runtime\org.eclipse.m2e.archetype.common\pom.xml
	at org.eclipse.tycho.core.utils.TychoProjectUtils.getTargetPlatformConfiguration(TychoProjectUtils.java:59)
	at org.eclipse.tycho.core.resolver.DefaultDependencyResolverFactory.lookupDependencyResolver(DefaultDependencyResolverFactory.java:43)
	at org.eclipse.tycho.target.TargetPlatformMojo.execute(TargetPlatformMojo.java:75)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	... 21 more
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :org.eclipse.m2e.archetype.common
```

and if that is handled second:

```
[INFO] --- tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath (default-validate-classpath) @ org.eclipse.m2e.archetype.common ---
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-compiler-jdt:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-compiler-jdt:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-compiler-jdt:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-compiler-jdt:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-lib-detector:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-lib-detector:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-lib-detector:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Skipped remote request for org.eclipse.tycho:tycho-lib-detector:2.7.0-SNAPSHOT/maven-metadata.xml locally installed metadata up-to-date
[DEBUG] Dependency collection stats {ConflictMarker.analyzeTime=32700, ConflictMarker.markTime=29700, ConflictMarker.nodeCount=38, ConflictIdSorter.graphTime=30500, ConflictIdSorter.topsortTime=15500, ConflictIdSorter.conflictIdCount=29, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=231600, ConflictResolver.conflictItemCount=38, DefaultDependencyCollector.collectTime=15025800, DefaultDependencyCollector.transformTime=359300}
[DEBUG] org.eclipse.tycho:tycho-compiler-plugin:jar:2.7.0-SNAPSHOT
[DEBUG]    org.codehaus.plexus:plexus-utils:jar:3.4.1:compile
[DEBUG]    org.codehaus.plexus:plexus-compiler-manager:jar:2.10.0:compile
[DEBUG]       org.codehaus.plexus:plexus-component-annotations:jar:2.1.1:compile (version managed from default)
[DEBUG]    org.codehaus.plexus:plexus-compiler-api:jar:2.10.0:compile
[DEBUG]    org.eclipse.tycho:tycho-core:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:org.eclipse.tycho.p2.tools.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:org.eclipse.tycho.p2.resolver.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.codehaus.plexus:plexus-archiver:jar:4.2.7:compile (version managed from default)
[DEBUG]          org.codehaus.plexus:plexus-io:jar:3.2.0:compile (version managed from default)
[DEBUG]             commons-io:commons-io:jar:2.6:compile
[DEBUG]          org.apache.commons:commons-compress:jar:1.21:compile (version managed from default)
[DEBUG]          org.iq80.snappy:snappy:jar:0.4:compile
[DEBUG]          org.tukaani:xz:jar:1.9:runtime
[DEBUG]       org.codehaus.plexus:plexus-interpolation:jar:1.26:compile
[DEBUG]       org.codehaus.plexus:plexus-cipher:jar:2.0:compile (version managed from default)
[DEBUG]          javax.inject:javax.inject:jar:1:compile
[DEBUG]       org.eclipse.tycho:sisu-equinox-embedder:jar:2.7.0-SNAPSHOT:compile
[DEBUG]          org.eclipse.platform:org.eclipse.osgi:jar:3.17.100:compile (version managed from default)
[DEBUG]          org.eclipse.tycho:sisu-equinox-api:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:tycho-metadata-model:jar:2.7.0-SNAPSHOT:compile
[DEBUG]          de.pdark:decentxml:jar:1.4:compile (version managed from default)
[DEBUG]       org.eclipse.tycho:tycho-embedder-api:jar:2.7.0-SNAPSHOT:compile
[DEBUG]          org.eclipse.tycho:org.eclipse.tycho.embedder.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:org.eclipse.tycho.core.shared:jar:2.7.0-SNAPSHOT:compile
[DEBUG]    org.eclipse.tycho:tycho-compiler-jdt:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.tycho:tycho-lib-detector:jar:2.7.0-SNAPSHOT:compile
[DEBUG]       org.eclipse.jdt:ecj:jar:3.28.0:compile (version managed from default)
[DEBUG]       org.apache.commons:commons-exec:jar:1.3:compile (version managed from default)
[DEBUG] Created new class realm plugin>org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT
[DEBUG] Importing foreign packages into class realm plugin>org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT
[DEBUG]   Imported:  < project>org.eclipse.m2e:m2e-maven-runtime:1.16.0-SNAPSHOT
[DEBUG] Populating class realm plugin>org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT
[DEBUG]   Included: org.eclipse.tycho:tycho-compiler-plugin:jar:2.7.0-SNAPSHOT
[DEBUG]   Included: org.codehaus.plexus:plexus-utils:jar:3.4.1
[DEBUG]   Included: org.codehaus.plexus:plexus-compiler-manager:jar:2.10.0
[DEBUG]   Included: org.codehaus.plexus:plexus-component-annotations:jar:2.1.1
[DEBUG]   Included: org.codehaus.plexus:plexus-compiler-api:jar:2.10.0
[DEBUG]   Included: org.codehaus.plexus:plexus-archiver:jar:4.2.7
[DEBUG]   Included: org.codehaus.plexus:plexus-io:jar:3.2.0
[DEBUG]   Included: commons-io:commons-io:jar:2.6
[DEBUG]   Included: org.apache.commons:commons-compress:jar:1.21
[DEBUG]   Included: org.iq80.snappy:snappy:jar:0.4
[DEBUG]   Included: org.tukaani:xz:jar:1.9
[DEBUG]   Included: org.codehaus.plexus:plexus-interpolation:jar:1.26
[DEBUG]   Included: org.codehaus.plexus:plexus-cipher:jar:2.0
[DEBUG]   Included: org.eclipse.platform:org.eclipse.osgi:jar:3.17.100
[DEBUG]   Included: org.eclipse.tycho:tycho-compiler-jdt:jar:2.7.0-SNAPSHOT
[DEBUG]   Included: org.eclipse.tycho:tycho-lib-detector:jar:2.7.0-SNAPSHOT
[DEBUG]   Included: org.eclipse.jdt:ecj:jar:3.28.0
[DEBUG]   Included: org.apache.commons:commons-exec:jar:1.3
[DEBUG] Configuring mojo org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath from plugin realm ClassRealm[plugin>org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@3764951d]
[DEBUG] Configuring mojo 'org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath' with basic configurator -->
[DEBUG]   (f) project = MavenProject: org.eclipse.m2e:org.eclipse.m2e.archetype.common:1.18.1-SNAPSHOT @ C:\dev\git\eclipse.m2e-core\m2e-maven-runtime\org.eclipse.m2e.archetype.common\pom.xml
[DEBUG]   (f) session = org.apache.maven.execution.MavenSession@39c7fb0b
[DEBUG] -- end configuration --
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] M2E - Maven runtime bundles 1.16.0-SNAPSHOT ........ SUCCESS [  1.069 s]
[INFO] M2E Maven Archetype Common 1.18.1-SNAPSHOT ......... FAILURE [  0.123 s]
[INFO] M2E Maven/Nexus Indexer 1.18.2-SNAPSHOT ............ SKIPPED
[INFO] M2E Embedded Maven Runtime (includes Incubating components) 1.18.2-SNAPSHOT SKIPPED
[INFO] M2E SLF4J-Simple Binding for Embedded Maven Runtime 1.18.1-SNAPSHOT SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.772 s
[INFO] Finished at: 2022-02-22T09:45:42+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath (default-validate-classpath) on project org.eclipse.m2e.archetype.common: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath failed: Project not setup correctly -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath (default-validate-classpath) on project org.eclipse.m2e.archetype.common: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath failed: Project not setup correctly
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:215)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:972)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:293)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:196)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath failed: Project not setup correctly
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:148)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	... 20 more
Caused by: java.lang.NullPointerException: Project not setup correctly
	at java.base/java.util.Objects.requireNonNull(Objects.java:246)
	at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getMavenSession(OsgiBundleProject.java:184)
	at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getBundleClassPath(OsgiBundleProject.java:393)
	at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getClasspath(OsgiBundleProject.java:380)
	at org.eclipse.tycho.compiler.ValidateClassPathMojo.execute(ValidateClassPathMojo.java:53)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	... 21 more
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :org.eclipse.m2e.archetype.common
```

> If both are expected to be skipped with `tycho.mode=maven` they could simply check this property in the `execute()` method.

That's what I have already done in this PR :)
However since this set-up is quite M2E specific and one can disable these plug-ins in their pom by binding them to the none phase, I would be fine if that property would not be checked in those plug-ins like it was done before. In other scenarios other plug-ins need to be disabled and other phases are executed so in the end almost every plug-in (at least those executed automatically would have to check that property). And with mixed reactor build tycho.mode=maven should become more and more obsolete. So I think it is ok to reduce the places where it is considered to the minimum.

For m2e I added the following in the `m2e-maven-runtime/pom.xml` to the pluginManagement section of the `generate-osgi-metadata` profile.
```
<plugin>
	<groupId>org.eclipse.tycho</groupId>
	<artifactId>target-platform-configuration</artifactId>
	<executions>
		<execution>
			<id>default-target-platform</id>
			<phase>none</phase>
		</execution>
	</executions>
</plugin>
<plugin>
	<groupId>org.eclipse.tycho</groupId>
	<artifactId>tycho-compiler-plugin</artifactId>
	<executions>
		<execution>
			<id>default-validate-classpath</id>
			<phase>none</phase>
		</execution>
	</executions>
</plugin>
```

laeubi commented:
> But how am I then supposed to handle situations in M2E where a first build in 'Maven-Mode' is done where I don't want the target-resolution to happen but in a subsequent build it should happen?

You should not use Tycho if you not like it to be effectively being used ;-)

So the first build should be either plain maven, or both builds should be merged into one where all projects that should not be build by tycho are using a jar or bundle-packaging.

> do you have any idea why both kick in?

tycho-build include tycho-pomless.

> Actually I would like to avoid the now 'legacy' extension tycho-pomless and move forward.

then you should move away from mode=maven :-)

> That's what I have already done in this PR :)

I think thats perfectly fine then.

laeubi commented:
> To be honest, I don't understand why you are fine with suppressing the execution of plug-ins when `tycho.mode=maven` is set but are against disabling the initial target resolution.

You can still sue that but *not* together with tycho-build plugin! So you can either use `tycho.mode=maven` with pom less-extension or you use the  tycho-build plugin but not both.

> should still work like it did for 2.6 and before

It did work 'as before' because before 2.7. there was no tycho-build-plugin ...

HannesWell commented:
> > To be honest, I don't understand why you are fine with suppressing the execution of plug-ins when `tycho.mode=maven` is set but are against disabling the initial target resolution.
> 
> You can still sue that but _not_ together with tycho-build plugin! So you can either use `tycho.mode=maven` with pom less-extension or you use the tycho-build plugin but not both.

OK if it is intended to work only with the pom-less extension overwriting the `tycho.mode` property still has to be removed from the `TychoPomlessLifecycleParticipant`. Otherwise the `tycho.mode` cannot be used even when using only `pom-less` and not `tycho-build`.

I have updated https://github.com/eclipse-m2e/m2e-core/pull/589 to stay on pom-less for now.
But this only works for 2.7.0 if `TychoPomlessLifecycleParticipant` does not overwrite `tycho.mode`.

laeubi reviewed (approved):
Looks fine, what happened to your changes regarding skip of Mojos? Is this no longer required?

HannesWell commented:
> Looks fine, what happened to your changes regarding skip of Mojos? Is this no longer required?

I disabled them in M2E's `generate-osgi-metadata` profile by binding the default executions to the `none` phase (you can see it in https://github.com/eclipse-m2e/m2e-core/pull/589). So it is at least not longer required to handle it in Tycho. Since the goal is to make tycho.mode obsolete we should not relay on it in more places.

laeubi commented:
I have cherry picke this to 2.7.x as well

HannesWell commented:
> I have cherry picke this to 2.7.x as well

👍🏽 Thanks!
